golang/cosmos/x/vstorage/keeper/keeper.go

```diff
@@ -129,7 +141,31 @@ func (k Keeper) ImportStorage(ctx sdk.Context, entries []*types.DataEntry) {
 	for _, entry := range entries {
 		// This set does the bookkeeping for us in case the entries aren't a
 		// complete tree.
-		k.SetStorage(ctx, entry.Path, entry.Value)
+		k.SetStorage(ctx, types.NewStorageEntry(entry.Path, entry.Value))
+	}
+}
+
+func (k Keeper) MigrateNoDataPlaceholders(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, nil)
+
+	// Copy empty keys first since cosmos stores do not support writing keys
+	// while an iterator is open over the domain
+	emptyKeys := []*([]byte){}
+	{
+		defer iterator.Close()
+		for ; iterator.Valid(); iterator.Next() {
+			rawValue := iterator.Value()
+			if bytes.Equal(rawValue, types.EncodedDataPrefix) {
+				key := iterator.Key()
+				emptyKeys = append(emptyKeys, &key)
+			}
+		}
+	}
```

Comment:
What is the purpose of this wrapping block? `defer` is function-scoped, so if you're looking to ensure that `iterator.Close()` is called before any `store.Set` then you'll need to refactor like e.g.
```suggestion
	readerClosed := false
	defer func() {
		if !readerClosed {
			iterator.Close()
		}
	}()
	for ; iterator.Valid(); iterator.Next() {
		rawValue := iterator.Value()
		if bytes.Equal(rawValue, types.EncodedDataPrefix) {
			key := iterator.Key()
			emptyKeys = append(emptyKeys, &key)
		}
	}
	iterator.Close()
	readerClosed = true
```
which can be simplified further if `iterator.Close()` is safe to call multiple times:
```suggestion
	defer iterator.Close()
	for ; iterator.Valid(); iterator.Next() {
		rawValue := iterator.Value()
		if bytes.Equal(rawValue, types.EncodedDataPrefix) {
			key := iterator.Key()
			emptyKeys = append(emptyKeys, &key)
		}
	}
	iterator.Close()
```

Reply:
Ugh what is the point of blocks if you can't scope defer to them...

Reply:
Given that it'd have to basically OOM panic during the iteration, I think I will simply remove the `defer` and add a single `iterator.Close` after the iteration is complete (which I actually assume automatically closes the iterator on its own, but hey, let's be good citizens)

Reply:
It looks like `sdk.KVStorePrefixIterator` is [cosmos-sdk `KVStorePrefixIterator`](https://github.com/cosmos/cosmos-sdk/blob/261af6721b84a45286bcdf812b0006d2d192f5c4/store/types/utils.go#L10-L13), which returns an implementation of [`KVStore` `Iterator`](https://github.com/cosmos/cosmos-sdk/blob/261af6721b84a45286bcdf812b0006d2d192f5c4/core/store/store.go#L19-L25) having "_Iterator must be closed by caller_" commentary. The Iterator interface itself is a reference to [github.com/cosmos/cosmos-db](https://github.com/cosmos/cosmos-db), specifically https://github.com/cosmos/cosmos-db/blob/313e4da3bdc5cf48f49300fac91549dc3ff1fc6d/types.go#L108-L157 , which expresses no requirements with respect to calling `Close` more than once.

Your plan of a single non-deferred `Close()` after exhausting the read iterator makes sense to me.